src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+	// is one which are not meant to survive the exit of the function they are created in.  The idea is that the caller
+	// knows the size of the buffer they need and they want to put the buffer on stack, if it doesn't exceed a certain
+	// value, or dynamically allocate memory if more is needed.  Even though the class could be used on its own, it's
+	// really meant to be a base for more specialized buffers.  There are not many safeguards, by design - the code is
+	// meant to be performant.  This is the reason why the size is static throughout the life of the object, so that we
+	// can perform as few checks as possible.
+
+	template <size_t MaxStackSize, typename T> using static_local_storage_base = local_storage<MaxStackSize, false, T>;
+
+	template<size_t MaxStackSize, typename T = char>
+	class static_local_storage final : public static_local_storage_base<MaxStackSize, T>
+	{
+		using base = static_local_storage_base<MaxStackSize, T>;
+
+	public:
+		explicit static_local_storage (size_t initial_size) noexcept
+			: base (initial_size)
+		{}
+	};
+
+	template <size_t MaxStackSize, typename T> using dynamic_local_storage_base = local_storage<MaxStackSize, true, T>;
+
+	template<size_t MaxStackSize, typename T = char>
+	class dynamic_local_storage final : public dynamic_local_storage_base<MaxStackSize, T>
+	{
+		using base = dynamic_local_storage_base<MaxStackSize, T>;
+
+	public:
+		explicit dynamic_local_storage (size_t initial_size = 0) noexcept
+			: base (initial_size)
+		{}
+
+		//
+		// If `new_size` is smaller than the current size and the dynamic store is allocated, data WILL NOT be
+		// preserved.
+		//
+		// If `new_size` is bigger than the current size bigger than MaxStackSize, data will be copied to the new
+		// dynamically allocated store
+		//
+		// If `new_size` is smaller or equal to `MaxStackSize`, no changes are made unless dynamic store was allocated,
+		// in which case it will be freed
+		//
+		void resize (size_t new_size) noexcept
+		{
+			size_t old_size = base::size ();
+
+			if (new_size == old_size) {
+				return;
+			}
+
+			if (new_size <= MaxStackSize) {
+				new_size = MaxStackSize;
+				base::free_store ();
+				return;
+			}
+
+			if (new_size < old_size) {
+				base::free_store ();
+				base::init_store (new_size);
+				return;
+			}
+
+			T* old_allocated_store = base::get_allocated_store ();
+			base::init_store (new_size);
+
+			T* new_allocated_store = base::get_allocated_store ();
+			if (old_allocated_store != nullptr) {
+				std::memcpy (new_allocated_store, old_allocated_store, old_size);
+				delete[] old_allocated_store;
+				return;
+			}
+
+			std::memcpy (new_allocated_store, base::get_local_store ().data (), MaxStackSize);
+		}
+	};
+
+	// This class mostly performs NO BOUNDS CHECKING - child classes MUST implement that!
+	template<size_t MaxStackSize, typename TStorage, typename TChar = char>
+	class string_base
```

Comment:
High-level, could we just have this be a closer mirror of `std::basic_string`?

Reply:
What do you mean? We have `assign`, `append`, `clear`, `starts_with` just like in `std::string`  with similar semantics, what else would you like to change to make it closer to `std::string`?